src/platform/ESP32/bluedroid/ChipDeviceScanner.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#if CONFIG_ENABLE_ESP32_BLE_CONTROLLER
+
+#include <platform/ESP32/ChipDeviceScanner.h>
+
+#include "esp_bt.h"
+#include "esp_bt_main.h"
+#include "esp_gap_ble_api.h"
+#include "esp_gatt_common_api.h"
+#include "esp_gatt_defs.h"
+#include "esp_gattc_api.h"
+#include "esp_gatts_api.h"
+#include "esp_log.h"
+#include "freertos/FreeRTOS.h"
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#define CHIPoBLE_SERVICE_UUID 0xFFF6
+
+namespace chip {
+namespace DeviceLayer {
+namespace Internal {
+namespace {
+
+/// Retrieve CHIP device identification info from the device advertising data
+bool BluedroidGetChipDeviceInfo(esp_ble_gap_cb_param_t & scan_result, chip::Ble::ChipBLEDeviceIdentificationInfo & deviceInfo)
+{
+    // Check for CHIP Service UUID
+    if (scan_result.scan_rst.ble_adv != NULL)
+    {
+        if (scan_result.scan_rst.adv_data_len > 13 && scan_result.scan_rst.ble_adv[5] == 0xf6 &&
+            scan_result.scan_rst.ble_adv[6] == 0xff)
+        {
+            deviceInfo.OpCode                              = scan_result.scan_rst.ble_adv[7];
+            deviceInfo.DeviceDiscriminatorAndAdvVersion[0] = scan_result.scan_rst.ble_adv[8];
+            deviceInfo.DeviceDiscriminatorAndAdvVersion[1] = scan_result.scan_rst.ble_adv[9];
+            // vendor and product Id from adv
+            deviceInfo.DeviceVendorId[0]  = scan_result.scan_rst.ble_adv[10];
+            deviceInfo.DeviceVendorId[1]  = scan_result.scan_rst.ble_adv[11];
+            deviceInfo.DeviceProductId[0] = scan_result.scan_rst.ble_adv[12];
+            deviceInfo.DeviceProductId[1] = scan_result.scan_rst.ble_adv[13];
+            deviceInfo.AdditionalDataFlag = scan_result.scan_rst.ble_adv[14];
+            return true;
+        }
```

Comment:
This does not correctly check first whether the advertisement field type is "Service Data - 16 bit UUID". It also assumes this field is at a specific position. From the matter spec: "_Note that the position of the fields within the Advertising PDU (e.g. Flags, 16-bit UUID Service Data, etc.) within a conformant advertisement MAY differ from the example above, since there are no ordering constraints for advertisement fields in the Length-Type-Value format used in BLE advertisements._". And at last, the length check "> 13" is wrong since it reads later at position 14.